mobius-core/src/main/java/com/spotify/mobius/DiscardAfterDisposeConnectable.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * -\-\-
+ * Mobius
+ * --
+ * Copyright (c) 2017-2018 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.mobius;
+
+import static com.spotify.mobius.DiscardAfterDisposeWrapper.wrapConnection;
+import static com.spotify.mobius.DiscardAfterDisposeWrapper.wrapConsumer;
+import static com.spotify.mobius.internal_util.Preconditions.checkNotNull;
+
+import com.spotify.mobius.disposables.CompositeDisposable;
+import com.spotify.mobius.disposables.Disposable;
+import com.spotify.mobius.functions.Consumer;
+import javax.annotation.Nonnull;
+
+/**
+ * A {@link Connectable} that ensures that an inner {@link Connection} doesn't emit or receive any
+ * values after being disposed.
+ *
+ * <p>This only acts as a safeguard, you still need to make sure that the Connectable disposes of
+ * resources correctly.
+ */
+class DiscardAfterDisposeConnectable<I, O> implements Connectable<I, O> {
+
+  private final Connectable<I, O> actual;
+
+  DiscardAfterDisposeConnectable(Connectable<I, O> actual) {
+    this.actual = checkNotNull(actual);
+  }
+
+  @Nonnull
+  @Override
+  public Connection<I> connect(Consumer<O> output) {
+    final DiscardAfterDisposeWrapper<O> safeOutput = wrapConsumer(output);
+    final Connection<I> input = actual.connect(safeOutput);
+    final DiscardAfterDisposeWrapper<I> safeInput = wrapConnection(input);
+
+    final Disposable disposable = CompositeDisposable.from(safeInput, safeOutput);
+
+    return new Connection<I>() {
+      @Override
+      public synchronized void accept(I effect) {
+        safeInput.accept(effect);
+      }
+
+      @Override
+      public synchronized void dispose() {
```

Comment:
```suggestion
      public void dispose() {
```
The wrapped method is `synchronized` already.